File: src/infra/captcha_service/hcaptcha.rs
use super::{CaptchaService, ValidateError};
use crate::domain::models::Captcha;
use async_trait::async_trait;
use serde::Deserialize;
use std::{collections::HashMap, env};

pub struct HCaptchaService {
    secret: String,
}

impl HCaptchaService {
    pub fn new() -> Self {
        Self {
            secret: env::var("HCAPTCHA_SECRET").expect("HCAPTCHA_SECRET env var"),
        }
    }

    fn form(&self, captcha: Captcha) -> HashMap<&'static str, String> {
        let mut form = HashMap::new();

        form.insert("response", captcha.into());
        form.insert("secret", self.secret.clone());

        form
    }
}

#[async_trait]
impl CaptchaService for HCaptchaService {
    async fn validate(&self, captcha: Captcha) -> Result<(), ValidateError> {
        let client = reqwest::Client::new();
        let res = match client
            .post("https://hcaptcha.com/siteverify")
            .form(&self.form(captcha))
            .send()
            .await
        {
            Ok(res) => res,
            _ => return Err(ValidateError::Unknown("could not send request")),
        };

        let hcaptcha_response = match res.json::<HCaptchaResponse>().await {
            Ok(res) => res,
            _ => return Err(ValidateError::Unknown("could not convert to domain")),
        };

        if hcaptcha_response.success {
            Ok(())
        } else {
            Err(ValidateError::Unknown("the captcha is invalid"))
        }
    }
}

#[derive(Deserialize)]
pub struct HCaptchaResponse {
    success: bool,
}

File: src/infra/captchaService/hcaptcha.ts
import { Env, Result } from "@utils";
import { Captcha } from "@domain";
import { CaptchaService } from "./mod.ts";
import { z } from "zod";

export class HCaptchaService implements CaptchaService {
    private secret: string;

    private constructor(secret: string) {
        this.secret = secret;
    }

    private form(captcha: Captcha) {
        const form = new FormData();
        form.append("response", Captcha.toString(captcha));
        form.append("secret", this.secret);
        return form;
    }

    static new() {
        return new HCaptchaService(
            Env.get("HCAPTCHA_SECRET")
        );
    }

    async validate(captcha: Captcha): Promise<Result<void, string>> {
        const response = await fetch("https://hcaptcha.com/siteverify", {
            method: "POST",
            body: this.form(captcha)
        });
        if (! response.ok) { return Result.err("HCaptchaService: could not send request"); }
        
        const hCaptchaResponse = hCaptchaResponseSchema
            .safeParse(await response.json());
        if (! hCaptchaResponse.success) { return Result.err("HCaptchaService: could not parse json"); }

        if (! hCaptchaResponse.data.success) { return Result.err("HCaptchaService: could not convert to domain"); }
        return Result.ok(undefined);
    }
}

const hCaptchaResponseSchema = z.object({
    success: z.boolean()
});